packages/lib/gpc/src/gpcCompile.ts

```diff
@@ -0,0 +1,738 @@
+  }[],
+  maxEntries: number
+): {
+  circuitEntryInputs: {
+    entryObjectIndex: CircuitSignal[];
+    entryNameHash: CircuitSignal[];
+    entryIsValueEnabled: CircuitSignal;
+    entryIsValueHashRevealed: CircuitSignal;
+  };
+  circuitEntryOutputs: {
+    entryRevealedValueHash: CircuitSignal[];
+  };
+} {
+  // Spare entry slots are filled with the name of Entry 0, value disabled.
+  for (
+    let entryIndex = allEntryInputsOutputs.length;
+    entryIndex < maxEntries;
+    entryIndex++
+  ) {
+    allEntryInputsOutputs.push({
+      inputs: {
+        objectIndex: allEntryInputsOutputs[0].inputs.objectIndex,
+        nameHash: allEntryInputsOutputs[0].inputs.nameHash,
+        isValueEnabled: 0n,
+        isValueHashRevealed: 0n
+      },
+      outputs: {
+        revealedValueHash: BABY_JUB_NEGATIVE_ONE
+      }
+    });
+  }
+
+  // Combine indidvidual arrays to form the circuit inputs, as 1D arrays, 2D
+  // arrays, or bitfields as appropriate.
+  return {
+    circuitEntryInputs: {
+      entryObjectIndex: allEntryInputsOutputs.map(
+        (io) => io.inputs.objectIndex
+      ),
+      entryNameHash: allEntryInputsOutputs.map((io) => io.inputs.nameHash),
+      entryIsValueEnabled: array2Bits(
+        allEntryInputsOutputs.map((io) => io.inputs.isValueEnabled)
+      ),
+      entryIsValueHashRevealed: array2Bits(
+        allEntryInputsOutputs.map((io) => io.inputs.isValueHashRevealed)
+      )
+    },
+    circuitEntryOutputs: {
+      entryRevealedValueHash: allEntryInputsOutputs.map(
+        (io) => io.outputs.revealedValueHash
+      )
+    }
+  };
+}
+
+function compileVerifyOwner(
+  ownerInput: GPCRevealedOwnerClaims | undefined,
+  firstOwnerIndex: number
+): {
+  circuitOwnerInputs: {
+    ownerEntryIndex: CircuitSignal;
+    ownerExternalNullifier: CircuitSignal;
+    ownerIsNullfierHashRevealed: CircuitSignal;
+  };
+  circuitOwnerOutputs: {
+    ownerRevealedNulifierHash: CircuitSignal;
+  };
+} {
+  // Owner module is enabled if any entry config declared it was an owner
+  // commitment.  It can't be enabled purely for purpose of nullifier hash,
+  // since an unconstrained owner could be set to any random numbers.
+  const hasOwner = firstOwnerIndex !== 0;
+
+  return {
+    circuitOwnerInputs: {
+      ownerEntryIndex: hasOwner
+        ? BigInt(firstOwnerIndex)
+        : BABY_JUB_NEGATIVE_ONE,
+      ownerExternalNullifier: makeWatermarkSignal(
+        ownerInput?.externalNullifier
+      ),
+      ownerIsNullfierHashRevealed:
+        ownerInput?.externalNullifier !== undefined ? 1n : 0n
+    },
+    circuitOwnerOutputs: {
+      ownerRevealedNulifierHash:
+        ownerInput?.nullifierHash ?? BABY_JUB_NEGATIVE_ONE
+    }
+  };
+}
+
+/**
+ * Creates a high-level description of the public claims of a proof, by
+ * redacting information from the proof's inputs and outputs.
+ *
+ * This code assumes that the arguments have already been checked to be
+ * well-formed represent a valid proof configuration using
```

Comment:
Same comment as above, this could use either punctuation, parentheses or something to make absolutely clear the relationship between "be well-formed" and "represent a valid proof configuration".